File: Algorithm/Java/198. House Robber.java
/**
* Author: LC
* Date:   2017-05-26 14:14:58
* Last modified by:   WuLC
* Last Modified time: 2017-05-26 14:34:35
* Email: liangchaowu5@gmail.com
*/

// dp
// dp[i] represents the max money that can robbed from with the first i houses
// dp[i] = max(dp[i-1], dp[i-2] + nums[i])
// compress it to O(1) space by representing dp[i-1] = curr and dp[i-1] = pre
public class Solution 
{
    public int rob(int[] nums) 
    {
        int curr = 0, pre = 0, tmp;
        for (int i = 0; i < nums.length; i++) 
        {
            tmp = curr;
            curr =Math.max(pre + nums[i], curr);
            pre = tmp;
        }
        return Math.max(curr, pre);
    }
}


// dp
// dp[i] represents the max money that can robbed from with the first i houses, and the i-th  house must be robbed
// dp[i] = max(dp[i-3], dp[i-2])+nums[i]
// O(n) space
public class Solution 
{
    public int rob(int[] nums) 
    {
        if (nums.length == 0) return 0;
        int n = nums.length;
        int[] dp = new int[n+1];
        for (int i = 0; i < n; i++) 
        {
            if (i < 2) dp[i+1] = nums[i];
            else dp[i+1] = Math.max(dp[i-1], dp[i-2]) + nums[i];
        }
        return Math.max(dp[n], dp[n-1]);
    }
}

File: Algorithm/Python/198. House Robber.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-06-11 13:19:29
# @Last modified by:   WuLC
# @Last Modified time: 2016-06-11 13:19:45
# @Email: liangchaowu5@gmail.com


# DP
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        nums.insert(0, 0)
        for i in xrange(3, len(nums)):
            nums[i] = max(nums[i]+nums[i-2], nums[i]+nums[i-3])
        return max(nums)

# dp in another way, when at certain index, robber may not rob this index
# constant space
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        p1, p2, p3 = 0, 0, 0
        for num in nums:
            tmp = max(p3, p2 + num)
            p1, p2 = p2, p3
            p3 = tmp
        return p3